Use JNA to Speed up Snapshot Cache File Creation

original-brownbear opened the pull request:
Use JNA to speed up snapshot cache file creation. Do this in `:server`
to bypass the security filter and move necessary bits of code to `:server`
to enable the logic.
Fall back to trying to create the file by writing zeros if anything except for
the step determining free disk space fails.

Manually verified via `du` that this works as expected and actually allocates hundreds of GB in a few `ms`.

original-brownbear committed:
Use JNA to Speed up Snapshot Cache File Creation

Use JNA to speed up snapshot cache file creation. Do this in `:server`
to bypass the security filter and move necessary bits of code to `:server`
to enable the logic.
Fall back to trying to create the file by writing zeros if anything except for
the step determining free disk space fails.

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

original-brownbear committed:
javadoc

original-brownbear committed:
javadoc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into jna-magic

original-brownbear committed:
bck

original-brownbear committed:
move things around a little nicer

original-brownbear commented on server/src/main/java/org/elasticsearch/bootstrap/JNANatives.java:
```diff
@@ -260,4 +269,39 @@ static void tryInstallSystemCallFilter(Path tmpFile) {
     }
+
+    @SuppressForbidden(reason = "need access to fd on FileOutputStream")
+    static void fallocateSnapshotCacheFile(Environment environment, long fileSize) throws IOException {
```
It's a little unfortunate to leak that much logic into this class, but I figured abstracting things away further to isolate the `fallocate` as much as possible wasn't really worth the extra complication (IMO).

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -132,6 +133,26 @@
 
     public static final String UPDATE_SNAPSHOT_STATUS_ACTION_NAME = "internal:cluster/snapshot/update_snapshot_status";
 
+    public static final String SEARCHABLE_SNAPSHOT_SETTINGS_PREFIX = "xpack.searchable.snapshot.shared_cache.";
```
this is the prefix for the shared cache, not for all SEARCHABLE_SNAPSHOT_SETTINGS

original-brownbear replied:
++ renamed.

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -132,6 +133,26 @@
 
+    public static final String SEARCHABLE_SNAPSHOT_SETTINGS_PREFIX = "xpack.searchable.snapshot.shared_cache.";
+
+    public static final Setting<ByteSizeValue> FROZEN_CACHE_RANGE_SIZE_SETTING = Setting.byteSizeSetting(
```
let's name this SHARED_CACHE_RANGE_SIZE_SETTING

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into jna-magic

original-brownbear committed:
CR: renamings

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -132,6 +133,26 @@
+        Setting.Property.NodeScope
+    );
+
+    public static final String CACHE_FILE_NAME = "snap_cache";
```
++ Renamed to `shared_snapshot_cache` :)

original-brownbear commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into jna-magic

original-brownbear committed:
Fix off by one Mistake in SparseFileTrackerTests (#68739)

This was caused by #68709 which turned the edge case of a
range with start > end into a tripped assertion.

ywelsch commented on server/src/main/java/org/elasticsearch/bootstrap/JNANatives.java:
```diff
@@ -260,4 +269,39 @@ static void tryInstallSystemCallFilter(Path tmpFile) {
+        }
+        Path cacheFile = SnapshotUtils.findCacheSnapshotCacheFilePath(environment, fileSize);
+        if (cacheFile == null) {
+            throw new IOError(new IOException("Could not find a directory with adequate free space for cache file"));
```
why wrap in IOError?

original-brownbear replied:
Remove that wrapping, sorry for the noise, this was a left-over from some debugging.

ywelsch reviewed (approved):
LGTM. Would like to see a review from one more person before merge

original-brownbear committed:
remove needless IOError

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into jna-magic

tlrx commented on server/src/main/java/org/elasticsearch/bootstrap/Natives.java:
```diff
@@ -132,4 +134,20 @@ static boolean isSystemCallFilterInstalled() {
+     */
+    public static void tryCreateCacheFile(Environment environment, long fileSize) throws IOException {
+        if (JNA_AVAILABLE == false) {
+            logger.warn("cannot use fallocate to create cache file because JNA is not available");
```
nit: I know we are not consistent in starting log messages with uppercase or not but we're using both ways in this PR

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -132,6 +133,26 @@
+    );
+    public static final Setting<ByteSizeValue> SNAPSHOT_CACHE_REGION_SIZE_SETTING = Setting.byteSizeSetting(
+        SHARED_CACHE_SETTINGS_PREFIX + "region_size",
+            SHARED_CACHE_RANGE_SIZE_SETTING,
```
nit ident

tlrx reviewed (approved):
LGTM

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into jna-magic

original-brownbear committed:
CR: nits

original-brownbear commented:
Thanks Yannick + Tanguy!

joegallo commented:
Reverted in 4c2e9a6d99849ab9970e18f747b7229a73ded5cc because it broke local development on a mac. (Sorry!)
